sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala

```diff
@@ -901,6 +901,91 @@ abstract class ParquetQuerySuite extends QueryTest with ParquetTest with SharedS
       }
     }
   }
+
+  test("SPARK-36634: Support access and read parquet file by column index") {
+    withTempDir { dir =>
+      val loc = s"file:///$dir/t"
+
+      withTable("t1", "t2", "t3") {
+        sql(s"create table t1 (my_id int, my_name string) using parquet location '$loc'")
+        sql(s"create table t2 (myid int, myName string) using parquet location '$loc'")
+        sql("insert into t1 select 1, 'apache'")
+        sql("insert into t2 select 2, 'software'")
+        sql("insert into t2 select 3, 'foundation'")
+        sql(s"create table t3 (myid int, myname string, myage int) using parquet location '$loc'")
+
+        withSQLConf((SQLConf.PARQUET_ACCESS_BY_ORDINAL.key, "false")) {
+          checkAnswer(sql("select my_id from t1"), Seq(Row(1), Row(null), Row(null)))
+          checkAnswer(sql("select my_id, my_name from t1"),
+            Seq(Row(1, "apache"), Row(null, null), Row(null, null)))
+          assert(sql("select my_id, my_name from t1 where my_id=2").isEmpty)
+          checkAnswer(sql("select myid, myname, myage from t3"),
+            Seq(Row(2, "software", null),
+              Row(3, "foundation", null),
+              Row(null, null, null)))
+        }
+
+        sql("insert into t3 select 4, 'spark', 11")
+
+        withAllParquetReaders {
+          withSQLConf((SQLConf.PARQUET_ACCESS_BY_ORDINAL.key, "true")) {
+            checkAnswer(sql("select my_id from t1"), Seq(Row(1), Row(2), Row(3)))
+            val e1 = {
+              intercept[SparkException](sql("select my_name from t1").collect())
+            }
+            assert(e1.getCause.getMessage.contains("Parquet column cannot be converted in"))
```

Comment:
hmm, so this feature breaks column pruning. How useful is this feature then?

Reply:
shall we turn off column pruning when access-by-ordinal is enabled?

Reply:
I guess they are not conflicting. We are using idx-0 to get data from `file 1(my_id)` and  `file 2(myid)` and feed it to my_name. Here the idx is not the final output set but the original input set of raw data

Reply:
Right now by-ordinal access only works if the query selects all the columns. Maybe the column pruning should pass "required column ordinals" to the parquet reader

Reply:
Not 'all the columns' but the resolved project list zipped directly to the fields of parquet data files

Reply:
I'm a bit confused. Say if a parquet file has 3 columns `a, b, c`. By-ordinal access only works if a query needs to read all the columns (no column pruning). Do I miss something?

Reply:
we can read `a`, `ab`, or `abc` by ordinal.

Reply:
But the test here shows we can't, or did I miss something?

Reply:
in this test, the user and the parser side requests `b`, but at the execution side we follow `by-ordinal` rule to read the parquet file, so `a` is read actually. If `a` has the same data type, there will be side effects; otherwise, we fail like this test

Reply:
I don't think that's how by-ordinal access should work. Let's say that the actual parquet schema is [a, b, c], and the catalyst schema is [x, y, z]. If a query reads column `y`, we should read column `b` from Parquet. It's very confusing to the users if `SELECT x`, `SELECT y` and `SELECT z` all return the same value, which is column `a` of the parquet file (assume all columns have the same type).

Reply:
Think about a user running `SELECT x, y` and `SELECT y`, he/she will see the column `y` has different values in each run, which is kind of a correctness bug.

Reply:
What's worse, turning off column pruning also changes the result, which definitely violates the principle (query optimization should not change result)

Reply:
I see. Let me think it over.